doc/reference/index.rst

```diff
@@ -23,3 +23,265 @@ API Reference
    usb/index.rst
    usermode/index.rst
    runtime_conf/index.rst
+
+Terminology
+***********
+
+The following terms may be used as shorthand API tags to indicate the
+allowed calling context (thread, ISR, pre-kernel), the effect of a call
+on the current thread state, and other behavioral characteristics.
+
+:ref:`api_term_reschedule`
+   if executing the function reaches a reschedule point
+:ref:`api_term_sleep`
+   if executing the function can cause the invoking thread to sleep
```

Comment:
This is fine from the perspective of API docs, though note that internally there is a distinction between sleep/suspend and "pend", which is a thread blocked on a wait_q.   Might be worth clarifying that here for the benefit of anyone trying to match this up to scheduler code.  Alternatively you could use the term "block" (which zephyr doesn't currently) as an umbrella term for both states.

Reply:
Yeah, it's a bit odd, but I was told "can't sleep" is the criteria for deciding when something could be used in interrupt context.  There's some background on choices and historical usage of different terms (including "block" which people tend to think means "synchronous") in https://github.com/pabigot/zephyr/blob/api/terms/doc/api18970.md.  In the end "sleep" was the least inaccurate candidate.